### ISSUE TITLE
Type constrain will cause the inferred type to be less specific

### ISSUE THREAD

**Issue by xbtsw** [microsoft/TypeScript#28422]
**TypeScript Version:**  3.2.0-dev.20181107

**Search Terms:**
Type constrain inferred
**Code**

```ts
function test1<P extends any[], R>(
  t: ((...args: P) => R),
): (...args: P) => R {
    return t;
}

function wrap1<T extends (...args: any[]) => any>(resultFn: T): T{
    return resultFn;
}

const test11 = wrap1(test1);

function wrap2<T>(resultFn: T): T{
    return resultFn;
}

const test12 = wrap2(test1);
```

**Expected behavior:**
`typeof test11` should be the same as `typeof test1` as the `wrap1` function takes `T` and return `T`.
**Actual behavior:**
`typeof test11` is more generic than `typeof test1`. Meanwhile, `wrap2` works as intended, except it doesn't have type constrain.

**Playground Link:** 
[Playground Link](https://www.typescriptlang.org/play/index.html#src=function%20test1%3CP%20extends%20any%5B%5D%2C%20R%3E(%0D%0A%20%20t%3A%20((...args%3A%20P)%20%3D%3E%20R)%2C%0D%0A)%3A%20(...args%3A%20P)%20%3D%3E%20R%20%7B%0D%0A%20%20%20%20return%20t%3B%0D%0A%7D%0D%0A%0D%0Afunction%20wrap1%3CT%20extends%20(...args%3A%20any%5B%5D)%20%3D%3E%20any%3E(resultFn%3A%20T)%3A%20T%7B%0D%0A%20%20%20%20return%20resultFn%3B%0D%0A%7D%0D%0A%0D%0Aconst%20test11%20%3D%20wrap1(test1)%3B%0D%0A%0D%0Afunction%20wrap2%3CT%3E(resultFn%3A%20T)%3A%20T%7B%0D%0A%20%20%20%20return%20resultFn%3B%0D%0A%7D%0D%0A%0D%0Aconst%20test12%20%3D%20wrap2(test1)%3B)

**Related Issues:** 
N/A

**Comment by RyanCavanaugh** [comment]
Either this is a bug *or* the logic for moving type parameters to the return type intentionally considers the constraint to be a request from the signature to effectively instantiate the generic function. Not sure which.

**Comment by ahejlsberg** [comment]
Fixed by #33228.
